A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,258 @@
+
+### Subchannel Wrapper
+
+The subchannel wrappers created by the `outlier_detection` LB policy will hold a reference to its map entry in the LB policy, if that map entry exists.
+
+The subchannel wrapper will track the latest state update from the underlying subchannel. By default, it will simply pass those updates along. The wrapper will have the following methods:
+
+ - `eject()`: The wrapper will report a state update with the `TRANSIENT_FAILURE` state, and will stop passing along updates from the underlying subchannel.
+ - `uneject()`: The wrapper will report a state update with the latest update from the underlying subchannel, and resume passing along updates from the underlying subchannel.
+
+### Success Rate Algorithm
+
+ 1. If the number of addresses with request volume of at least `success_rate_ejection.request_volume` is less than `success_rate_ejection.minimum_hosts`, stop.
+ 2. Calculate the mean and standard deviation of the fractions of successful requests among addresses with total request volume of at least `success_rate_ejection.request_volume`.
+ 3. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `success_rate_ejection.request_volume`, continue to the next address.
+    3. If the address's success rate is less than `(mean - stdev * (success_rate_ejection.stdev_factor / 1000))`, then choose a random integer in `[0, 100)`. If that number is less than `success_rate_ejection.enforcement_percentage`, eject that address.
+
+### Failure Percentage Algorithm
+
+ 1. If the number of addresses is less than `failure_percentage_ejection.minimum_hosts`, stop.
+ 2. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `failure_percentage_ejection.request_volume`, continue to the next address.
+    3. If the address's failure percentage is greater than `failure_percentage_ejection.threshold`, then choose a random integer in `[0, 100)`. If that number is less than `failiure_percentage_ejection.enforcement_percentage`, eject that address.
+
+### xDS Integration
+
+If the `Cluster` message for an underlying cluster has the `outlier_detection` field set, the `xds_cluster_resolver` will generate an `outlier_detection` LB policy config from it for each priority as the top-level LB policy for that priority. In that `outlier_detection` field message, if the `enforcement_percentage` field is set to 0, the config `success_rate_ejection` field will be `null` and all `success_rate_*` fields will be ignored. If the `enforcing_failure_percent` field is set to 0 or `null`, the config `failure_percent_ejection` field will be `null` and all `failure_percent_*` fields will be ignored. Then the message fields will be mapped to config fields as follows:
```

Comment:
Since we've decided to always add the `outlier_detection` policy in the tree when we're using xDS, we should remove the words "If the Cluster message for an underying cluster has the outlier_detection field set", since that makes it sound conditional.

Reply:
It is conditional, in the sense that if the field is not set, the procedure described here is not followed, and the no-op config is generated instead. That's described in the next paragraph, but that's after the list so I can understand if it doesn't seem connected.

Reply:
Ah, I see.  Yeah, it's a bit hard to follow that way for a first-time reader.  I think it would be clearer to say something like this:

"""
The `xds_cluster_resolver` LB policy will configure an `outlier_detection` LB policy instance for each priority as the top-level LB policy for that priority.  The configuration for that `outlier_detection` policy will be based on the `outlier_detection` field in the `Cluster` resource.

If the `outlier_detection` field is not set in the `Cluster` resource, a "no-op" `outlier_detection` config will be generated, with `interval` set to the maximum possible value and all other fields unset.

If the `outlier_detection` field *is* set in the `Cluster` resource, if the `enforcement_percentage` field is set to 0, the config `success_rate_ejection` field will be `null` and all `success_rate_*` fields will be ignored. If the `enforcing_failure_percent` field is set to 0 or `null`, the config `failure_percent_ejection` field will be `null` and all `failure_percent_*` fields will be ignored. Then the message fields will be mapped to config fields as follows:
"""